configuration.md

# Topcoder X Configuration

The following config parameters are supported, they are defined in `src/config.js` and can be configured as env variables:


| Name | Description | Default |
| :------------------------------------- | :----------------------------------------: | :------------------------------: |
| PORT                                   | the port the application will listen on | 80                              |
| API_VERSION | the API version | v1 |
| LOG_LEVEL                              | the log level | info                            |
| SESSION_SECRET | the session secret | kjsdfkj34857 |
| GITHUB_CLIENT_ID                       | the GitHub client id | No default - needs to be set up using the instructions below                                 |
| GITHUB_CLIENT_SECRET | the GitHub client secret | No default - needs to be set up using the instructions below |
| GITLAB_CLIENT_ID                       | the GitLab client id | No default - needs to be set up using the instructions below                                                           |
| GITLAB_CLIENT_SECRET | the GitLab client secret | No default - needs to be set up using the instructions below |
| WEBSITE                                | used as base to construct various URLs | http://topcoderx.topcoder-dev.com/ |
| GITLAB_API_BASE_URL | The Gitlab API base URL |  https://gitlab.com|
|TOPIC  | The Kafka topic where events are published.  This must be the same as the configured value for topcoder-x-processor| |
|KAFKA_OPTIONS | Kafka connection options|  |
|KAFKA_URL | The Kafka host to connect to| localhost:9092 |
|KAFKA_CLIENT_CERT | The Kafka SSL certificate to use when connecting| Read from kafka_client.cer file, but this can be set as a string like it is on Heroku |
|KAFKA_CLIENT_CERT_KEY | The Kafka SSL certificate key to use when connecting| Read from kafka_client.key file, but this can be set as a string like it is on Heroku|
| HOOK_BASE_URL | The base URL of the topcoder-x-receiver, used when adding webhooks automatically to repositories |  |
| TOPCODER_ENV | The topcoder environment to use, can support 'dev' or 'prod' | 'dev'                     |
|LABELS| Labels we are going to add to the repository in the form of array of object with `name` and `color` property. Color should be hex code without hash||
|ALLOWED_TOPCODER_ROLES| The allowed Topcoder role to use Topcoder X app| see configuration |
|COPILOT_ROLE| The role to identify copilot|'copilot'|
|HELP_LINK| The link for help| 'https://github.com/topcoder-platform/topcoder-x-ui/wiki'|
|ADMINISTRATOR_ROLES| The array of roles to be considered as admin|  `['administrator', 'admin']`|
|AWS_ACCESS_KEY_ID | The Amazon certificate key to use when connecting. Use local dynamodb you can set fake value|FAKE_ACCESS_KEY_ID |
|AWS_SECRET_ACCESS_KEY | The Amazon certificate access key to use when connecting. Use local dynamodb you can set fake value|FAKE_SECRET_ACCESS_KEY |
|AWS_REGION | The Amazon certificate region to use when connecting. Use local dynamodb you can set fake value|FAKE_REGION |
|IS_LOCAL | Use Amazon DynamoDB Local or server. |true |
|AWS_CONNECTION_TIMEOUT | The timeout used to check if the app is healthy. |10000 |
|TC_LOGIN_URL | TC login url | |
|DYNAMODB_WAIT_TABLE_FOR_ACTIVE_TIMEOUT | Dynamodb wait for active timeout |10 minutes |
|TC_API_V5_URL | Topcoder API v5 url for retrieving list of Connect Projects | |


## GitHub OAuth App Setup

These instructions should be used to generate the GITHUB_CLIENT_ID and GITHUB_CLIENT_SECRET values for configuration

- Login into github.com
- Click the upper right avatar, then click `Settings`
- Click the left panel --> Developer settings --> OAuth Apps
- Click the `Register a new application`, fill in the fields,
  note that the `Authorization callback URL` should be the deployed topcoder-x web site,
  for local deployment, it should be `http://topcoderx.topcoder-dev.com`
- After creating the OAuth app, you can see its client id and client secret,
  these should be set to GITHUB_CLIENT_ID and GITHUB_CLIENT_SECRET environment variables

## GitLab OAuth App Setup

These instructions should be used to generate the GITLAB_CLIENT_ID and GITLAB_CLIENT_SECRET values for configuration


- Login into gitlab.com
- Click the upper right avatar, then click `Settings`
- Click the `Applications` tab
- Enter an application name, e.g. `Topcoder-X`
- For Redirect URI, enter two callback URLs, one callback URL per line, so there are two lines:
  http://topcoderx.topcoder-dev.com/api/v1/gitlab/owneruser/callback
 http://topcoderx.topcoder-dev.com/api/v1/gitlab/normaluser/callback
- For Scopes, check the `api` and `read_user`, the `api` is for owner user, the `read_user` is for normal user
- Finally, click `Save application` to save the OAuth app, then you will see its generated Application Id and Secret,
  these should be set to GITLAB_CLIENT_ID and GITLAB_CLIENT_SECRET environment variables

  ## Topcoder environment

  Application depends upon the environment of Topcoder based on TOPCODER_ENV configuration. When this value is 'dev' you should be using topcoderx.topcoder-dev.com.

  To switch to production, simply change the TOPCODER_DEV to 'prod' and use topcoderx.topcoder.com

  Default configuration is for topcoder-dev environment.


src/config.js
/*
 * Copyright (c) 2017 TopCoder, Inc. All rights reserved.
 */
const fs = require('fs');
/**
 * Define config.
 *
 * @author TCSCODER
 * @version 1.0
 */
module.exports = {
  PORT: process.env.PORT || 80, // eslint-disable-line no-magic-numbers
  API_VERSION: process.env.API_VERSION || 'v1',
  LOG_LEVEL: process.env.LOG_LEVEL || 'debug',
  SESSION_SECRET: process.env.SESSION_SECRET || 'kjsdfkj34857',
  // Github and gitlab client id and secret
  GITHUB_CLIENT_ID: process.env.GITHUB_CLIENT_ID,
  GITHUB_CLIENT_SECRET: process.env.GITHUB_CLIENT_SECRET,
  GITLAB_CLIENT_ID: process.env.GITLAB_CLIENT_ID,
  GITLAB_CLIENT_SECRET: process.env.GITLAB_CLIENT_SECRET,

  // used as base to construct various URLs
  WEBSITE: process.env.WEBSITE || 'http://topcoderx.topcoder-dev.com',
  GITLAB_API_BASE_URL: process.env.GITLAB_API_BASE_URL || 'https://gitlab.com',

  // kafka configuration
  TOPIC: process.env.TOPIC || 'tc-x-events',
  KAFKA_OPTIONS: {
    connectionString: process.env.KAFKA_URL || 'localhost:9092',
    ssl: {
      cert: process.env.KAFKA_CLIENT_CERT || fs.readFileSync('./kafka_client.cer'), // eslint-disable-line no-sync
      key: process.env.KAFKA_CLIENT_CERT_KEY || fs.readFileSync('./kafka_client.key'), // eslint-disable-line no-sync
      passphrase: 'secret', // NOTE:* This configuration specifies the private key passphrase used while creating it.
    },
  },
  HOOK_BASE_URL: process.env.HOOK_BASE_URL || 'http://topcoderx.topcoder-dev.com',
  TOPCODER_ENV: process.env.TOPCODER_ENV || 'dev',
  LABELS: process.env.LABELS || [
    {name: 'tcx_OpenForPickup', color: '428BCA'},
    {name: 'tcx_Assigned', color: '004E00'},
    {name: 'tcx_ReadyForReview', color: 'D1D100'},
    {name: 'tcx_Paid', color: '7F8C8D'},
    {name: 'tcx_Feedback', color: 'FF0000'},
    {name: 'tcx_FixAccepted', color: '69D100'},
    {name: 'tcx_NotReady', color: '000000'},
    {name: 'tcx_Canceled', color: '000000'},
  ],
  OPEN_FOR_PICKUP_ISSUE_LABEL: process.env.OPEN_FOR_PICKUP_ISSUE_LABEL || 'tcx_OpenForPickup',
  ALLOWED_TOPCODER_ROLES: process.env.ALLOWED_TOPCODER_ROLES || ['administrator', 'admin', 'connect manager', 'connect admin', 'copilot', 'connect copilot'],
  COPILOT_ROLE: process.env.COPILOT_ROLE || 'copilot',
  ADMINISTRATOR_ROLES: process.env.ADMINISTRATOR_ROLES || ['administrator', 'admin'],
  DYNAMODB: {
    AWS_ACCESS_KEY_ID: process.env.AWS_ACCESS_KEY_ID,
    AWS_SECRET_ACCESS_KEY: process.env.AWS_SECRET_ACCESS_KEY,
    AWS_REGION: process.env.AWS_REGION,
    IS_LOCAL: process.env.IS_LOCAL,
    TIMEOUT: process.env.AWS_CONNECTION_TIMEOUT || 10000, // eslint-disable-line no-magic-numbers
  },
  TOPCODER_VALUES: {
    dev: {
      TC_LOGIN_URL: process.env.TC_LOGIN_URL || 'https://accounts-auth0.topcoder-dev.com/',
    },
    prod: {
      TC_LOGIN_URL: process.env.TC_LOGIN_URL || 'https://accounts-auth0.topcoder.com/',
    },
  },
  DYNAMODB_WAIT_TABLE_FOR_ACTIVE_TIMEOUT: process.env.DYNAMODB_WAIT_TABLE_FOR_ACTIVE_TIMEOUT || 1000 * 60 * 10 // eslint-disable-line no-magic-numbers
};

module.exports.frontendConfigs = {
  copilotRole: module.exports.COPILOT_ROLE,
  administratorRoles: module.exports.ADMINISTRATOR_ROLES,
  helpLink: process.env.HELP_LINK || 'https://github.com/topcoder-platform/topcoder-x-ui/wiki',
  JWT_V3_NAME: process.env.JWT_V3_NAME || 'v3jwt',
  JWT_V2_NAME: process.env.JWT_V2_NAME || 'tcjwt',
  COOKIES_SECURE: process.env.COOKIES_SECURE || false,
  TC_LOGIN_URL: process.env.TC_LOGIN_URL || 'https://accounts-auth0.topcoder-dev.com/',
  ADMIN_TOOL_URL: process.env.ADMIN_TOOL_URL || 'https://api.topcoder-dev.com/v2',
  ACCOUNTS_CONNECTOR_URL: process.env.ACCOUNTS_CONNECTOR_URL || 'https://accounts.topcoder-dev.com/connector.html',
  CONNECT_URL_BASE: process.env.CONNECT_URL_BASE || 'https://connect.topcoder-dev.com/projects/',
  OWNER_LOGIN_GITHUB_URL: process.env.OWNER_LOGIN_GITHUB_URL || '/api/v1/github/owneruser/login',
  OWNER_LOGIN_GITLAB_URL: process.env.OWNER_LOGIN_GITLAB_URL || '/api/v1/gitlab/owneruser/login',
  TOPCODER_URL: process.env.TOPCODER_URL || 'https://topcoder-dev.com',
  GITHUB_TEAM_URL: process.env.GITHUB_TEAM_URL || 'https://github.com/orgs/',
  GITLAB_GROUP_URL: process.env.GITLAB_GROUP_URL || 'https://gitlab.com/groups/',
  TC_API_V5_URL: process.env.TC_API_V5_URL || 'https://api.topcoder.com/v5',
  TOPCODER_VALUES: {
    dev: {
      TC_API_V4_URL: process.env.TC_API_V4_URL || 'https://api.topcoder-dev.com/v4',
    },
    prod: {
      TC_API_V4_URL: process.env.TC_API_V4_URL || 'https://api.topcoder.com/v4',
    },
  },
  TOPCODER_ENV: process.env.TOPCODER_ENV || 'prod',
};
